### PR TITLE
Convert byte_extract parser to Rust.

### PR DESCRIPTION
**Pull request by jlucovsky**
Convert the byte_extract option parser from C to Rust.

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [6873](https://redmine.openinfosecfoundation.org/issues/6873)

Describe changes:
- Refactor code in rust/src/detect to support re-usability
- Implement the byte_extract parser in Rust, with unittests
- Modify the existing byte_extract module to use the Rust parser

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 19262

File: rust/src/detect/byte_extract.rs [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+
+#[derive(Debug)]
+enum ResultValue {
+    Numeric(u64),
+    String(String),
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct DetectByteExtractData {
```
Comment (jasonish) [inline rust/src/detect/byte_extract.rs:60]: Looks like this is exposed to C, so should have the `SC` prefix.

File: rust/src/detect/byte_extract.rs [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+            "string and slice are mutually exclusive".to_string(),
+        ));
+    }
+
+    Ok((input, byte_extract))
+}
+
+/// Intermediary function between the C code and the parsing functions.
+#[no_mangle]
+pub unsafe extern "C" fn ScByteExtractParse(c_arg: *const c_char) -> *mut DetectByteExtractData {
```
Comment (jasonish) [inline rust/src/detect/byte_extract.rs:299]: `SC`, not `Sc`.

File: src/detect-byte-extract.c [hunk 3]
```diff
@@ -38,27 +34,24 @@
...
 
 #include "util-byte.h"
 #include "util-debug.h"
 #include "util-unittest.h"
 #include "util-unittest-helper.h"
-#include "util-spm.h"
 
 /* the default value of endianness to be used, if none's specified */
-#define DETECT_BYTE_EXTRACT_ENDIAN_DEFAULT DETECT_BYTE_EXTRACT_ENDIAN_BIG
+#define DETECT_BYTE_EXTRACT_ENDIAN_DEFAULT (uint8_t) BigEndian
```
Comment (jasonish) [inline src/detect-byte-extract.c:47]: `BigEndian` is coming from the Rust enum right?

File: src/detect-byte-extract.c [hunk 4]
```diff
@@ -38,27 +34,24 @@
...
 
 /* the base to be used if string mode is specified.  These options would be
  * specified in DetectByteParseData->base */
 #define DETECT_BYTE_EXTRACT_BASE_NONE 0
-#define DETECT_BYTE_EXTRACT_BASE_HEX  16
-#define DETECT_BYTE_EXTRACT_BASE_DEC  10
-#define DETECT_BYTE_EXTRACT_BASE_OCT   8
+#define DETECT_BYTE_EXTRACT_BASE_HEX  (uint8_t) BaseHex
```
Comment (jasonish) [inline src/detect-byte-extract.c:52]: From a Rust enum as well?  I've played with giving this a prefix, but it means the Rust names and the C names are different, breaking stuff like git grep, but it also puts these names into the export C namespace which isn't good either.  But really outside the scope of this PR. Just really seeing it now though.

File: src/detect-byte-extract.c (hunk 3 above)
Comment (jlucovsky) [reply src/detect-byte-extract.c:47]: Correct -- `rust/src/detect/lib.rs`

File: src/detect-byte-extract.c (hunk 4 above)
Comment (jlucovsky) [reply src/detect-byte-extract.c:52]: Yes --- also from `rust/src/detect/lib.rs`

File: rust/src/detect/byte_extract.rs (hunk 1 above)
Comment (jlucovsky) [reply rust/src/detect/byte_extract.rs:60]: Thanks. I'll fixup the byte math equivalents in a separate commit (also).

Does the naming prefix extend to flags exposed from Rust?

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 19332

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .tcp.pseudo | 2810 | 19562 | 696.16% | 

Pipeline 19629

File: rust/src/detect/byte_math.rs [hunk 5]
```diff
@@ -19,6 +19,7 @@
 
 use crate::detect::error::RuleParseError;
 use crate::detect::parser::{parse_token, take_until_whitespace};
+use crate::detect::*;
```
Comment (catenacyber) [inline rust/src/detect/byte_math.rs:22]: Is it not better to specify the precise list than having a wildcard ?

File: rust/src/detect/mod.rs [hunk 6]
```diff
@@ -39,7 +41,50 @@ pub trait EnumString<T> {
...
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ByteBase {
+    _BaseNone = 0,
+    BaseOct = 8,
+    BaseDec = 10,
+    BaseHex = 16,
+}
+
+fn get_string_value(value: &str) -> Result<ByteBase, ()> {
```
Comment (catenacyber) [inline rust/src/detect/mod.rs:69]: If we are refactoring, could this be an `Option<ByteBase>` ?  or implement trait FromStr ?

File: rust/src/detect/mod.rs [hunk 7]
```diff
@@ -65,6 +110,9 @@ mod test {
         assert_eq!(TestEnum::BestValueEver.to_str(), "best_value_ever");
         assert_eq!(TestEnum::from_str("zero"), Some(TestEnum::Zero));
         assert_eq!(TestEnum::from_str("nope"), None);
-        assert_eq!(TestEnum::from_str("best_value_ever"), Some(TestEnum::BestValueEver));
+        assert_eq!(
+            TestEnum::from_str("best_value_ever"),
```
Comment (catenacyber) [inline rust/src/detect/mod.rs:115]: Looks like `rustfmt` work that should be in its own commit

File: src/detect-byte-extract.h [hunk 8]
```diff
@@ -24,40 +24,11 @@
...
 
-/* flags */
-#define DETECT_BYTE_EXTRACT_FLAG_RELATIVE   0x01
-#define DETECT_BYTE_EXTRACT_FLAG_MULTIPLIER 0x02
-#define DETECT_BYTE_EXTRACT_FLAG_STRING     0x04
-#define DETECT_BYTE_EXTRACT_FLAG_ALIGN      0x08
-#define DETECT_BYTE_EXTRACT_FLAG_ENDIAN     0x10
-
 /* endian value to be used.  Would be stored in DetectByteParseData->endian */
 #define DETECT_BYTE_EXTRACT_ENDIAN_NONE    0
```
Comment (catenacyber) [inline src/detect-byte-extract.h:28]: you can remove this `DETECT_BYTE_EXTRACT_ENDIAN_NONE` as it is unused in C

File: rust/src/detect/byte_extract.rs [hunk 9]
```diff
@@ -0,0 +1,502 @@
+/* Copyright (C) 2022 Open Information Security Foundation
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:1]: copyright year ?

File: rust/src/detect/byte_extract.rs [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+            base: BASE_DEFAULT,                         // From string or dce
+            align_value: 0,
+            multiplier_value: DETECT_BYTE_EXTRACT_MULTIPLIER_DEFAULT,
+            id: 0,
+        }
+    }
+}
+
+impl SCDetectByteExtractData {
+    pub fn new() -> Self {
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:101]: Can we get rid of `new` and just use `default` ?

File: rust/src/detect/byte_extract.rs [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+    let value = values[1]
+        .parse::<i32>()
+        .map_err(|_| make_error(format!("invalid offset value: {}", values[1])))?;
+    if value >= i16::MIN.into() && value <= i16::MAX.into() {
+        byte_extract.offset = value as i16;
+    } else {
+        return Err(make_error(format!(
+            "invalid offset value: must be between {} and {}: {}",
+            u16::MIN,
+            u16::MAX,
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:150]: u16 or i16 ? error message and test disagree

File: rust/src/detect/byte_extract.rs [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+                        "base specified before string; use \"string, base\"".to_string(),
+                    ));
+                }
+                byte_extract.flags |= DETECT_BYTE_EXTRACT_FLAG_STRING;
+            }
+            "oct" | "dec" | "hex" => {
+                if 0 == (byte_extract.flags & DETECT_BYTE_EXTRACT_FLAG_STRING) {
+                    return Err(make_error("string must be set first".to_string()));
+                }
+                byte_extract.base = match get_string_value(name) {
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:228]: Should you test if a base was already set from DETECT_BYTE_EXTRACT_FLAG_BASE ?

File: rust/src/detect/byte_extract.rs [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+                    .map_err(|_| make_error(format!("invalid multiplier value: {}", val)))?;
+                if mult == 0 || mult > u16::MAX.into()
+                {
+                    return Err(make_error(format!(
+                        "invalid multiplier value: must be between 0 and {}: {}",
+                        u16::MAX,
+                        val
+                    )));
+                }
+                byte_extract.multiplier_value = mult as u16;
```
Comment (catenacyber) [inline rust/src/detect/byte_extract.rs:257]: Why parse a u32 and use a u16 ?

File: src/detect-byte-extract.c [hunk 14]
```diff
@@ -109,26 +89,22 @@ void DetectByteExtractRegister(void)
...
-
     if (payload_len == 0) {
         return 0;
     }
 
     /* Calculate the ptr value for the bytetest and length remaining in
      * the packet from that point.
      */
+    const uint8_t *ptr;
+    int32_t len;
```
Comment (catenacyber) [inline src/detect-byte-extract.c:106]: Why is it not initialized to 0 anymore ?

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks, looks good overall, some nits though

File: rust/src/detect/byte_math.rs (hunk 5 above)
Comment (jlucovsky) [reply rust/src/detect/byte_math.rs:22]: Good point; I'll make an update.

File: rust/src/detect/byte_extract.rs (hunk 13 above)
Comment (jlucovsky) [reply rust/src/detect/byte_extract.rs:257]: The value is limited to non-zero integer 2^16-1 so storing it as a `u16` made sense.

Is there a downside or upside to doing it differently?

File: rust/src/detect/mod.rs (hunk 7 above)
Comment (jlucovsky) [reply rust/src/detect/mod.rs:115]: If you feel strongly that it should be in a separate commit, I can make the update. There are a few other rustfmt changes in addition to this one.

File: src/detect-byte-extract.c (hunk 14 above)
Comment (jlucovsky) [reply src/detect-byte-extract.c:106]: Because either path through the `if` logic will assign a value.

File: rust/src/detect/byte_extract.rs (hunk 13 above)
Comment (catenacyber) [reply rust/src/detect/byte_extract.rs:257]: Why not parse directly as a u16 then ?

File: rust/src/detect/mod.rs (hunk 7 above)
Comment (catenacyber) [reply rust/src/detect/mod.rs:115]: Makes reviewing easier

**Comment by jlucovsky** [comment]
Continued in #10706
